pkg/internal/testing/addr/manager.go

```diff
@@ -28,33 +32,68 @@ import (
 const (
 	portReserveTime   = 1 * time.Minute
 	portConflictRetry = 100
+	portFilePrefix    = "port-"
 )
 
+var (
+	cacheDir string
+)
+
+func init() {
+	baseDir, err := os.UserCacheDir()
+	if err != nil {
+		baseDir = os.TempDir()
+	}
+	cacheDir = filepath.Join(baseDir, "kubebuilder-envtest")
+	_ = os.MkdirAll(cacheDir, 0750)
+}
+
 type portCache struct {
-	lock  sync.Mutex
-	ports map[int]time.Time
+	lock sync.Mutex
 }
 
-func (c *portCache) add(port int) bool {
+func (c *portCache) add(port int) (bool, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	// remove outdated port
-	for p, t := range c.ports {
-		if time.Since(t) > portReserveTime {
-			delete(c.ports, p)
+	// Remove outdated ports.
+	if err := fs.WalkDir(os.DirFS(cacheDir), ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
 		}
+		if d.IsDir() || !d.Type().IsRegular() || !strings.HasPrefix(path, portFilePrefix) {
+			return nil
+		}
+		info, err := d.Info()
+		if err != nil {
+			return err
+		}
+		if time.Since(info.ModTime()) > portReserveTime {
+			if err := os.Remove(filepath.Join(cacheDir, path)); err != nil {
+				panic(err)
```

Comment:
the func returns an error, why not just return the err?

Reply:
Oops, forgot during testing